apps/rush-lib/src/logic/DependencySpecifier.ts

```diff
@@ -44,6 +45,16 @@ export class DependencySpecifier {
 
   public constructor(packageName: string, versionSpecifier: string) {
     this.packageName = packageName;
+
+    // Workspace ranges are a feature from PNPM and Yarn. Set the version specifier
+    // to the trimmed version range.
+    if (versionSpecifier.startsWith('workspace:')) {
```

Comment:
The `workspace:` specifier is an interesting feature. Rush's equivalent is:

- If the dependency is SemVer-compatible with a Rush project version, it gets locally linked.
- You can disable that using `cyclicDependencyProjects`.

Issue https://github.com/microsoft/rushstack/issues/547 proposes to report an error if a dependency is NOT SemVer-compatible  and NOT in `cyclicDependencyProjects`.  And to rename it to `installedDependencyProjects`.

If we implemented that, then adding to `installedDependencyProjects` would be equivalent to NOT using the `workspace:` specifier.  We could potentially retire the `cyclicDependencyProjects`/`installedDependencyProjects` feature entirely and just use `workspace:` specifier.

But there's one downside of `workspace`:  In a monorepo, you *almost always* want to locally link projects.  Seems like  it would be very common beginner mistake to forget to add `workspace:` and then be confused why Rush is trying to install your package from the registry.

Maybe you guys could debate this and update https://github.com/microsoft/rushstack/issues/547 with your decision.

Reply:
Correct, that would be confusing :) I actually have a remedy for that problem in the workspace manager. When encountering this scenario, we check the package.json versions using the same logic for local-linking. If we determine that it would have been locally linked, then we modify the package.json file to specify `workspace:*` and write a warning to the console. If the version range doesn't match the local project, we validate that it's specified in `cyclicDependencyProjects`. If it is still not found, then we throw an error stating that it needs to either be added to `cyclicDependencyProjects`, or needs to target the local version.

In this way, the only time that a local package can be specified to be downloaded from the registry instead of locally linked would be to specify a normal version range in the package.json, ex `~1.2.3`, and add it to the `cyclicDependencyProjects` field.

Devs can also manually specify their own `workspace:~1.2.3` range for a local project. However, we will explicitly allow PNPM to handle resolution of that version range and let PNPM fail if it cannot be found.

Reply:
> If we determine that it would have been locally linked, then we modify the package.json file to specify `workspace:*` and write a warning to the console.

If you make it an error instead of a warning, I think you could resolve https://github.com/microsoft/rushstack/issues/547 with your PR.  

But then `cyclicDependencyProjects` is not just for cyclic dependencies any more; it is for any situation where you want to install a previously published release. So it would make sense to rename the setting to something like `installedDependencyProjects` or `nonlinkedProjects` or somesuch.  

To avoid breaking compatibility, we could document `cyclicDependencyProjects` as deprecated (removed in the next major release) and introduce the new setting alongside the old one. We could also prohibit `cyclicDependencyProjects` when `pnpmOptions.useWorkspaces = true`